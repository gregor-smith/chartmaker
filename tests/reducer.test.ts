import { reducer, Action } from '@/reducer'

import { createTestState } from './utils'
import { State } from '@/types'


type ActionParams = [ Action ]


const dispatchMock = jest.fn<void, ActionParams>()
afterEach(() => dispatchMock.mockReset())


const state = createTestState()


test('UpdateAPIKey', () => {
    const apiKey = 'New API key'
    const result = reducer(
        createTestState(),
        { tag: 'UpdateAPIKey', apiKey }
    )
    expect(result).toMatchSnapshot()
})


test('UpdateActiveChart', () => {
    const activeChartIndex = 123
    const result = reducer(
        createTestState(),
        {
            tag: 'UpdateActiveChart',
            index: activeChartIndex
        }
    )
    expect(result).toMatchSnapshot()
})


describe('PromptForNewChart', () => {
    // jsdom doesn't implement prompt so jest.spyOn can't be used here
    const promptMock = jest.fn<ReturnType<typeof prompt>, Parameters<typeof prompt>>()
    beforeAll(() => global.prompt = promptMock)
    afterEach(() => promptMock.mockRestore())
    afterAll(() => delete global.prompt)

    test('cancelling the prompt dispatches nothing', () => {
        promptMock.mockImplementation(() => null)

        const result = reducer(state, { tag: 'PromptForNewChart' })
        expect(result).toMatchSnapshot()

        const { sideEffect } = result as Extract<typeof result, { tag: 'SideEffect' }>
        sideEffect(dispatchMock, state)

        expect(dispatchMock).toHaveBeenCalledTimes(0)
    })

    test('entering nothing in the prompt dispatches nothing', () => {
        promptMock.mockImplementation(() => '')

        const result = reducer(state, { tag: 'PromptForNewChart' })
        expect(result).toMatchSnapshot()

        const { sideEffect } = result as Extract<typeof result, { tag: 'SideEffect' }>
        sideEffect(dispatchMock, state)

        expect(dispatchMock).toHaveBeenCalledTimes(0)
    })

    test('entering name of existing chart dispatches name taken action', () => {
        promptMock.mockImplementation(() => 'Test chart 1')

        const result = reducer(state, { tag: 'PromptForNewChart' })
        expect(result).toMatchSnapshot()

        const { sideEffect } = result as Extract<typeof result, { tag: 'SideEffect' }>
        sideEffect(dispatchMock, state)

        expect(dispatchMock).toHaveBeenCalledTimes(1)
        expect(dispatchMock).toHaveBeenCalledWith<ActionParams>({
            tag: 'ShowChartNameTakenMessage'
        })
    })

    test('entering unique name dispatches new chart action', () => {
        const name = 'Test new chart'
        promptMock.mockImplementation(() => name)

        const result = reducer(state, { tag: 'PromptForNewChart' })
        expect(result).toMatchSnapshot()

        const { sideEffect } = result as Extract<typeof result, { tag: 'SideEffect' }>
        sideEffect(dispatchMock, state)

        expect(dispatchMock).toHaveBeenCalledTimes(1)
        expect(dispatchMock).toHaveBeenCalledWith<ActionParams>({
            tag: 'AddNewChart',
            name
        })
    })
})


describe('ShowChartNameTakenMessage', () => {
    // likewise, it doesn't implement alert
    const alertMock = jest.fn<void, [ string ]>()
    beforeAll(() => global.alert = alertMock)
    afterEach(() => alertMock.mockRestore())
    afterAll(() => delete global.alert)

    test('calls alert', () => {
        const result = reducer(state, { tag: 'ShowChartNameTakenMessage' })
        expect(result).toMatchSnapshot()

        const { sideEffect } = result as Extract<typeof result, { tag: 'SideEffect' }>
        sideEffect(dispatchMock, state)

        expect(dispatchMock).toHaveBeenCalledTimes(0)
        expect(alertMock).toHaveBeenCalledTimes(1)
        expect(alertMock).toHaveBeenCalledWith('A chart with that name already exists')
    })
})


test('AddNewChart', () => {
    const result = reducer(
        state,
        {
            tag: 'AddNewChart',
            name: 'Test new chart'
        }
    )
    expect(result).toMatchSnapshot()
})


describe('PromptToRenameActiveChart', () => {
    const promptMock = jest.fn<ReturnType<typeof prompt>, Parameters<typeof prompt>>()
    beforeAll(() => global.prompt = promptMock)
    afterEach(() => promptMock.mockRestore())
    afterAll(() => delete global.prompt)

    test('cancelling the prompt dispatches nothing', () => {
        promptMock.mockImplementation(() => null)

        const result = reducer(state, { tag: 'PromptToRenameActiveChart' })
        expect(result).toMatchSnapshot()

        const { sideEffect } = result as Extract<typeof result, { tag: 'SideEffect' }>
        sideEffect(dispatchMock, state)

        expect(dispatchMock).toHaveBeenCalledTimes(0)
    })

    test('entering nothing in the prompt dispatches nothing', () => {
        promptMock.mockImplementation(() => '')

        const result = reducer(state, { tag: 'PromptToRenameActiveChart' })
        expect(result).toMatchSnapshot()

        const { sideEffect } = result as Extract<typeof result, { tag: 'SideEffect' }>
        sideEffect(dispatchMock, state)

        expect(dispatchMock).toHaveBeenCalledTimes(0)
    })

    test('entering the same name as active chart in the prompt dispatches nothing', () => {
        promptMock.mockImplementation(() => state.charts[state.activeChartIndex].name)

        const result = reducer(state, { tag: 'PromptToRenameActiveChart' })
        expect(result).toMatchSnapshot()

        const { sideEffect } = result as Extract<typeof result, { tag: 'SideEffect' }>
        sideEffect(dispatchMock, state)

        expect(dispatchMock).toHaveBeenCalledTimes(0)
    })

    test('entering name of other existing chart dispatches name taken action', () => {
        const name = 'Other test chart'
        promptMock.mockImplementation(() => name)

        const stateWithExtraChart: State = {
            ...state,
            charts: [
                ...state.charts,
                { ...state.charts[0], name }
            ]
        }
        const result = reducer(
            stateWithExtraChart,
            { tag: 'PromptToRenameActiveChart' }
        )
        expect(result).toMatchSnapshot()

        const { sideEffect } = result as Extract<typeof result, { tag: 'SideEffect' }>
        sideEffect(dispatchMock, stateWithExtraChart)

        expect(dispatchMock).toHaveBeenCalledTimes(1)
        expect(dispatchMock).toHaveBeenCalledWith<ActionParams>({
            tag: 'ShowChartNameTakenMessage'
        })
    })

    test('entering unique name dispatches rename action', () => {
        const name = 'Renamed test chart'
        promptMock.mockImplementation(() => name)

        const result = reducer(state, { tag: 'PromptToRenameActiveChart' })
        expect(result).toMatchSnapshot()

        const { sideEffect } = result as Extract<typeof result, { tag: 'SideEffect' }>
        sideEffect(dispatchMock, state)

        expect(dispatchMock).toHaveBeenCalledTimes(1)
        expect(dispatchMock).toHaveBeenCalledWith<ActionParams>({
            tag: 'RenameActiveChart',
            name
        })
    })
})


test.each<State>([
    state,
    createTestState({ charts: 2 }),
    {
        ...createTestState({ charts: 3 }),
        activeChartIndex: 2
    }
])('RenameActiveChart', state => {
    const name = 'Renamed test chart'
    const result = reducer(state, { tag: 'RenameActiveChart', name })
    expect(result).toMatchSnapshot()
})


describe('PromptToDeleteActiveChart', () => {
    const confirmMock = jest.fn<boolean, [ string | undefined ]>()
    beforeAll(() => global.confirm = confirmMock)
    afterEach(() => confirmMock.mockRestore())
    afterAll(() => delete global.confirm)

    test('declining the prompt dispatches nothing', () => {
        confirmMock.mockImplementation(() => false)

        const result = reducer(state, { tag: 'PromptToDeleteActiveChart' })
        expect(result).toMatchSnapshot()

        const { sideEffect } = result as Extract<typeof result, { tag: 'SideEffect' }>
        sideEffect(dispatchMock, state)

        expect(confirmMock).toHaveBeenCalledTimes(1)
        expect(confirmMock).toHaveBeenCalledWith('Really delete active chart? This cannot be undone')
        expect(dispatchMock).toHaveBeenCalledTimes(0)
    })

    test('accepting the prompt dispatches delete action', () => {
        confirmMock.mockImplementation(() => true)

        const result = reducer(state, { tag: 'PromptToDeleteActiveChart' })
        expect(result).toMatchSnapshot()

        const { sideEffect } = result as Extract<typeof result, { tag: 'SideEffect' }>
        sideEffect(dispatchMock, state)

        expect(confirmMock).toHaveBeenCalledTimes(1)
        expect(confirmMock).toHaveBeenCalledWith('Really delete active chart? This cannot be undone')
        expect(dispatchMock).toHaveBeenCalledTimes(1)
        expect(dispatchMock).toHaveBeenCalledWith<ActionParams>({
            tag: 'DeleteActiveChart'
        })
    })
})


describe('DeleteActiveChart', () => {
    test.each<State>([
        state,
        createTestState({ charts: 3 })
    ])('replaces active chart if only one chart, otherwise removes', state => {
        const result = reducer(state, { tag: 'DeleteActiveChart' })
        expect(result).toMatchSnapshot()
    })
})


describe('ImportStateFile', () => {
    test.todo('read error dispatches error action')


    test.todo('json parse error dispatches error action')


    test.todo('validation error dispatches error action')


    test.todo('validation success dispatches load state action')
})


test.todo('ShowInvalidStateImportMessage')


test.todo('LoadState')


test.todo('PromptToExportState')


describe('CancelSearchRequest', () => {
    test.todo('no update wehn search request not in progress')


    test.todo('changes search state to waiting and aborts request controller')
})


describe('SendSearchRequest', () => {
    test.todo('no update when request already in progress')


    test.todo('no update when search query empty')


    test.todo('error when api key empty')


    test.todo('side effect dispatches load state action on request success')


    test.todo('side effect dispatches error action on request status error')


    test.todo('side effect dispatches error action on request response json decode error')


    test.todo('side effect dispatches error action on request response validation error')


    test.todo('side effect dispatches error action on request network error')
})


test.todo('UpdateSearchState')


describe('UpdateSearchQuery', () => {
    test.todo('no update when search state loading')


    test.todo('update for other search states')
})


describe('DragChartAlbum', () => {
    test.todo('no update when source and target ids are same')


    test.todo('no update when album with source id cannot be found')


    test.todo('no update when album with target id cannot be found')


    test.todo('inserts source before target when source index higher')


    test.todo('inserts source after target when target index higher')
})


describe('DropSearchAlbum', () => {
    test.todo('no update when search state not complete')


    test.todo('no update when album with source id cannot be found')


    test.todo('no update when album with target id cannot be found')


    test.todo('replaces album at target index')
})


describe('PromptToRenameAlbum', () => {
    test.todo('dispatches nothing when album with id not found')


    test.todo('dispatches nothing if album with id is a placeholder')


    test.todo('dispatches nothing when prompt cancelled')


    test.todo('dispatches nothing when nothing entered in prompt')


    test.todo('dispatches rename action')
})


describe('RenameAlbum', () => {
    test.todo('no update when album with id not found')


    test.todo('no update when album with id is a placeholder')


    test.todo('renames album with id')
})


describe('DeleteAlbum', () => {
    test.todo('no update when album with id not found')


    test.todo('replaces album with id with a placeholder')
})


test.todo('UpdateScreenshotLoading')


describe('UpdateScreenshotScale', () => {
    test.todo('no update when screenshot in progress')


    test.todo('updates screenshot scale')
})


describe('TakeScreenshot', () => {
    test.todo('no update when screenshot in progress')


    test.todo('side effect downloads picture and dispatches action')
})


test.todo('UpdateChartShape')
