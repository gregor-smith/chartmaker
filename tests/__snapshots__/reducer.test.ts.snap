// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNewChart 1`] = `
Object {
  "state": Object {
    "activeChartIndex": 1,
    "albumIDCounter": 103,
    "apiKey": "Test API key",
    "charts": Array [
      Object {
        "albums": Array [
          Object {
            "id": 1,
            "placeholder": true,
          },
          Object {
            "id": 2,
            "placeholder": true,
          },
          Object {
            "id": 3,
            "placeholder": true,
          },
        ],
        "name": "Test chart 1",
        "rowsX": 10,
        "rowsY": 10,
        "shape": Object {
          "size": 40,
          "tag": "Top",
        },
      },
      Object {
        "albums": Array [
          Object {
            "id": 4,
            "placeholder": true,
          },
          Object {
            "id": 5,
            "placeholder": true,
          },
          Object {
            "id": 6,
            "placeholder": true,
          },
          Object {
            "id": 7,
            "placeholder": true,
          },
          Object {
            "id": 8,
            "placeholder": true,
          },
          Object {
            "id": 9,
            "placeholder": true,
          },
          Object {
            "id": 10,
            "placeholder": true,
          },
          Object {
            "id": 11,
            "placeholder": true,
          },
          Object {
            "id": 12,
            "placeholder": true,
          },
          Object {
            "id": 13,
            "placeholder": true,
          },
          Object {
            "id": 14,
            "placeholder": true,
          },
          Object {
            "id": 15,
            "placeholder": true,
          },
          Object {
            "id": 16,
            "placeholder": true,
          },
          Object {
            "id": 17,
            "placeholder": true,
          },
          Object {
            "id": 18,
            "placeholder": true,
          },
          Object {
            "id": 19,
            "placeholder": true,
          },
          Object {
            "id": 20,
            "placeholder": true,
          },
          Object {
            "id": 21,
            "placeholder": true,
          },
          Object {
            "id": 22,
            "placeholder": true,
          },
          Object {
            "id": 23,
            "placeholder": true,
          },
          Object {
            "id": 24,
            "placeholder": true,
          },
          Object {
            "id": 25,
            "placeholder": true,
          },
          Object {
            "id": 26,
            "placeholder": true,
          },
          Object {
            "id": 27,
            "placeholder": true,
          },
          Object {
            "id": 28,
            "placeholder": true,
          },
          Object {
            "id": 29,
            "placeholder": true,
          },
          Object {
            "id": 30,
            "placeholder": true,
          },
          Object {
            "id": 31,
            "placeholder": true,
          },
          Object {
            "id": 32,
            "placeholder": true,
          },
          Object {
            "id": 33,
            "placeholder": true,
          },
          Object {
            "id": 34,
            "placeholder": true,
          },
          Object {
            "id": 35,
            "placeholder": true,
          },
          Object {
            "id": 36,
            "placeholder": true,
          },
          Object {
            "id": 37,
            "placeholder": true,
          },
          Object {
            "id": 38,
            "placeholder": true,
          },
          Object {
            "id": 39,
            "placeholder": true,
          },
          Object {
            "id": 40,
            "placeholder": true,
          },
          Object {
            "id": 41,
            "placeholder": true,
          },
          Object {
            "id": 42,
            "placeholder": true,
          },
          Object {
            "id": 43,
            "placeholder": true,
          },
          Object {
            "id": 44,
            "placeholder": true,
          },
          Object {
            "id": 45,
            "placeholder": true,
          },
          Object {
            "id": 46,
            "placeholder": true,
          },
          Object {
            "id": 47,
            "placeholder": true,
          },
          Object {
            "id": 48,
            "placeholder": true,
          },
          Object {
            "id": 49,
            "placeholder": true,
          },
          Object {
            "id": 50,
            "placeholder": true,
          },
          Object {
            "id": 51,
            "placeholder": true,
          },
          Object {
            "id": 52,
            "placeholder": true,
          },
          Object {
            "id": 53,
            "placeholder": true,
          },
          Object {
            "id": 54,
            "placeholder": true,
          },
          Object {
            "id": 55,
            "placeholder": true,
          },
          Object {
            "id": 56,
            "placeholder": true,
          },
          Object {
            "id": 57,
            "placeholder": true,
          },
          Object {
            "id": 58,
            "placeholder": true,
          },
          Object {
            "id": 59,
            "placeholder": true,
          },
          Object {
            "id": 60,
            "placeholder": true,
          },
          Object {
            "id": 61,
            "placeholder": true,
          },
          Object {
            "id": 62,
            "placeholder": true,
          },
          Object {
            "id": 63,
            "placeholder": true,
          },
          Object {
            "id": 64,
            "placeholder": true,
          },
          Object {
            "id": 65,
            "placeholder": true,
          },
          Object {
            "id": 66,
            "placeholder": true,
          },
          Object {
            "id": 67,
            "placeholder": true,
          },
          Object {
            "id": 68,
            "placeholder": true,
          },
          Object {
            "id": 69,
            "placeholder": true,
          },
          Object {
            "id": 70,
            "placeholder": true,
          },
          Object {
            "id": 71,
            "placeholder": true,
          },
          Object {
            "id": 72,
            "placeholder": true,
          },
          Object {
            "id": 73,
            "placeholder": true,
          },
          Object {
            "id": 74,
            "placeholder": true,
          },
          Object {
            "id": 75,
            "placeholder": true,
          },
          Object {
            "id": 76,
            "placeholder": true,
          },
          Object {
            "id": 77,
            "placeholder": true,
          },
          Object {
            "id": 78,
            "placeholder": true,
          },
          Object {
            "id": 79,
            "placeholder": true,
          },
          Object {
            "id": 80,
            "placeholder": true,
          },
          Object {
            "id": 81,
            "placeholder": true,
          },
          Object {
            "id": 82,
            "placeholder": true,
          },
          Object {
            "id": 83,
            "placeholder": true,
          },
          Object {
            "id": 84,
            "placeholder": true,
          },
          Object {
            "id": 85,
            "placeholder": true,
          },
          Object {
            "id": 86,
            "placeholder": true,
          },
          Object {
            "id": 87,
            "placeholder": true,
          },
          Object {
            "id": 88,
            "placeholder": true,
          },
          Object {
            "id": 89,
            "placeholder": true,
          },
          Object {
            "id": 90,
            "placeholder": true,
          },
          Object {
            "id": 91,
            "placeholder": true,
          },
          Object {
            "id": 92,
            "placeholder": true,
          },
          Object {
            "id": 93,
            "placeholder": true,
          },
          Object {
            "id": 94,
            "placeholder": true,
          },
          Object {
            "id": 95,
            "placeholder": true,
          },
          Object {
            "id": 96,
            "placeholder": true,
          },
          Object {
            "id": 97,
            "placeholder": true,
          },
          Object {
            "id": 98,
            "placeholder": true,
          },
          Object {
            "id": 99,
            "placeholder": true,
          },
          Object {
            "id": 100,
            "placeholder": true,
          },
          Object {
            "id": 101,
            "placeholder": true,
          },
          Object {
            "id": 102,
            "placeholder": true,
          },
          Object {
            "id": 103,
            "placeholder": true,
          },
        ],
        "name": "Test new chart",
        "rowsX": 5,
        "rowsY": 5,
        "shape": Object {
          "size": 40,
          "tag": "Top",
        },
      },
    ],
    "screenshot": Object {
      "loading": false,
      "scale": 2,
    },
    "search": Object {
      "query": "Test query",
      "tag": "Waiting",
    },
  },
  "tag": "Update",
}
`;

exports[`DeleteActiveChart replaces active chart if only one chart, otherwise removes 1`] = `
Object {
  "state": Object {
    "activeChartIndex": 0,
    "albumIDCounter": 103,
    "apiKey": "Test API key",
    "charts": Array [
      Object {
        "albums": Array [
          Object {
            "id": 4,
            "placeholder": true,
          },
          Object {
            "id": 5,
            "placeholder": true,
          },
          Object {
            "id": 6,
            "placeholder": true,
          },
          Object {
            "id": 7,
            "placeholder": true,
          },
          Object {
            "id": 8,
            "placeholder": true,
          },
          Object {
            "id": 9,
            "placeholder": true,
          },
          Object {
            "id": 10,
            "placeholder": true,
          },
          Object {
            "id": 11,
            "placeholder": true,
          },
          Object {
            "id": 12,
            "placeholder": true,
          },
          Object {
            "id": 13,
            "placeholder": true,
          },
          Object {
            "id": 14,
            "placeholder": true,
          },
          Object {
            "id": 15,
            "placeholder": true,
          },
          Object {
            "id": 16,
            "placeholder": true,
          },
          Object {
            "id": 17,
            "placeholder": true,
          },
          Object {
            "id": 18,
            "placeholder": true,
          },
          Object {
            "id": 19,
            "placeholder": true,
          },
          Object {
            "id": 20,
            "placeholder": true,
          },
          Object {
            "id": 21,
            "placeholder": true,
          },
          Object {
            "id": 22,
            "placeholder": true,
          },
          Object {
            "id": 23,
            "placeholder": true,
          },
          Object {
            "id": 24,
            "placeholder": true,
          },
          Object {
            "id": 25,
            "placeholder": true,
          },
          Object {
            "id": 26,
            "placeholder": true,
          },
          Object {
            "id": 27,
            "placeholder": true,
          },
          Object {
            "id": 28,
            "placeholder": true,
          },
          Object {
            "id": 29,
            "placeholder": true,
          },
          Object {
            "id": 30,
            "placeholder": true,
          },
          Object {
            "id": 31,
            "placeholder": true,
          },
          Object {
            "id": 32,
            "placeholder": true,
          },
          Object {
            "id": 33,
            "placeholder": true,
          },
          Object {
            "id": 34,
            "placeholder": true,
          },
          Object {
            "id": 35,
            "placeholder": true,
          },
          Object {
            "id": 36,
            "placeholder": true,
          },
          Object {
            "id": 37,
            "placeholder": true,
          },
          Object {
            "id": 38,
            "placeholder": true,
          },
          Object {
            "id": 39,
            "placeholder": true,
          },
          Object {
            "id": 40,
            "placeholder": true,
          },
          Object {
            "id": 41,
            "placeholder": true,
          },
          Object {
            "id": 42,
            "placeholder": true,
          },
          Object {
            "id": 43,
            "placeholder": true,
          },
          Object {
            "id": 44,
            "placeholder": true,
          },
          Object {
            "id": 45,
            "placeholder": true,
          },
          Object {
            "id": 46,
            "placeholder": true,
          },
          Object {
            "id": 47,
            "placeholder": true,
          },
          Object {
            "id": 48,
            "placeholder": true,
          },
          Object {
            "id": 49,
            "placeholder": true,
          },
          Object {
            "id": 50,
            "placeholder": true,
          },
          Object {
            "id": 51,
            "placeholder": true,
          },
          Object {
            "id": 52,
            "placeholder": true,
          },
          Object {
            "id": 53,
            "placeholder": true,
          },
          Object {
            "id": 54,
            "placeholder": true,
          },
          Object {
            "id": 55,
            "placeholder": true,
          },
          Object {
            "id": 56,
            "placeholder": true,
          },
          Object {
            "id": 57,
            "placeholder": true,
          },
          Object {
            "id": 58,
            "placeholder": true,
          },
          Object {
            "id": 59,
            "placeholder": true,
          },
          Object {
            "id": 60,
            "placeholder": true,
          },
          Object {
            "id": 61,
            "placeholder": true,
          },
          Object {
            "id": 62,
            "placeholder": true,
          },
          Object {
            "id": 63,
            "placeholder": true,
          },
          Object {
            "id": 64,
            "placeholder": true,
          },
          Object {
            "id": 65,
            "placeholder": true,
          },
          Object {
            "id": 66,
            "placeholder": true,
          },
          Object {
            "id": 67,
            "placeholder": true,
          },
          Object {
            "id": 68,
            "placeholder": true,
          },
          Object {
            "id": 69,
            "placeholder": true,
          },
          Object {
            "id": 70,
            "placeholder": true,
          },
          Object {
            "id": 71,
            "placeholder": true,
          },
          Object {
            "id": 72,
            "placeholder": true,
          },
          Object {
            "id": 73,
            "placeholder": true,
          },
          Object {
            "id": 74,
            "placeholder": true,
          },
          Object {
            "id": 75,
            "placeholder": true,
          },
          Object {
            "id": 76,
            "placeholder": true,
          },
          Object {
            "id": 77,
            "placeholder": true,
          },
          Object {
            "id": 78,
            "placeholder": true,
          },
          Object {
            "id": 79,
            "placeholder": true,
          },
          Object {
            "id": 80,
            "placeholder": true,
          },
          Object {
            "id": 81,
            "placeholder": true,
          },
          Object {
            "id": 82,
            "placeholder": true,
          },
          Object {
            "id": 83,
            "placeholder": true,
          },
          Object {
            "id": 84,
            "placeholder": true,
          },
          Object {
            "id": 85,
            "placeholder": true,
          },
          Object {
            "id": 86,
            "placeholder": true,
          },
          Object {
            "id": 87,
            "placeholder": true,
          },
          Object {
            "id": 88,
            "placeholder": true,
          },
          Object {
            "id": 89,
            "placeholder": true,
          },
          Object {
            "id": 90,
            "placeholder": true,
          },
          Object {
            "id": 91,
            "placeholder": true,
          },
          Object {
            "id": 92,
            "placeholder": true,
          },
          Object {
            "id": 93,
            "placeholder": true,
          },
          Object {
            "id": 94,
            "placeholder": true,
          },
          Object {
            "id": 95,
            "placeholder": true,
          },
          Object {
            "id": 96,
            "placeholder": true,
          },
          Object {
            "id": 97,
            "placeholder": true,
          },
          Object {
            "id": 98,
            "placeholder": true,
          },
          Object {
            "id": 99,
            "placeholder": true,
          },
          Object {
            "id": 100,
            "placeholder": true,
          },
          Object {
            "id": 101,
            "placeholder": true,
          },
          Object {
            "id": 102,
            "placeholder": true,
          },
          Object {
            "id": 103,
            "placeholder": true,
          },
        ],
        "name": "Untitled chart",
        "rowsX": 5,
        "rowsY": 5,
        "shape": Object {
          "size": 40,
          "tag": "Top",
        },
      },
    ],
    "screenshot": Object {
      "loading": false,
      "scale": 2,
    },
    "search": Object {
      "query": "Test query",
      "tag": "Waiting",
    },
  },
  "tag": "Update",
}
`;

exports[`DeleteActiveChart replaces active chart if only one chart, otherwise removes 2`] = `
Object {
  "state": Object {
    "activeChartIndex": 1,
    "albumIDCounter": 3,
    "apiKey": "Test API key",
    "charts": Array [
      Object {
        "albums": Array [
          Object {
            "id": 4,
            "placeholder": true,
          },
          Object {
            "id": 5,
            "placeholder": true,
          },
          Object {
            "id": 6,
            "placeholder": true,
          },
        ],
        "name": "Test chart 2",
        "rowsX": 10,
        "rowsY": 10,
        "shape": Object {
          "size": 40,
          "tag": "Top",
        },
      },
      Object {
        "albums": Array [
          Object {
            "id": 7,
            "placeholder": true,
          },
          Object {
            "id": 8,
            "placeholder": true,
          },
          Object {
            "id": 9,
            "placeholder": true,
          },
        ],
        "name": "Test chart 3",
        "rowsX": 10,
        "rowsY": 10,
        "shape": Object {
          "size": 40,
          "tag": "Top",
        },
      },
    ],
    "screenshot": Object {
      "loading": false,
      "scale": 2,
    },
    "search": Object {
      "query": "Test query",
      "tag": "Waiting",
    },
  },
  "tag": "Update",
}
`;

exports[`PromptForNewChart cancelling the prompt dispatches nothing 1`] = `
Object {
  "sideEffect": [Function],
  "tag": "SideEffect",
}
`;

exports[`PromptForNewChart entering name of existing chart dispatches name taken action 1`] = `
Object {
  "sideEffect": [Function],
  "tag": "SideEffect",
}
`;

exports[`PromptForNewChart entering nothing in the prompt dispatches nothing 1`] = `
Object {
  "sideEffect": [Function],
  "tag": "SideEffect",
}
`;

exports[`PromptForNewChart entering unique name dispatches new chart action 1`] = `
Object {
  "sideEffect": [Function],
  "tag": "SideEffect",
}
`;

exports[`PromptToDeleteActiveChart accepting the prompt dispatches delete action 1`] = `
Object {
  "sideEffect": [Function],
  "tag": "SideEffect",
}
`;

exports[`PromptToDeleteActiveChart declining the prompt dispatches nothing 1`] = `
Object {
  "sideEffect": [Function],
  "tag": "SideEffect",
}
`;

exports[`PromptToRenameActiveChart cancelling the prompt dispatches nothing 1`] = `
Object {
  "sideEffect": [Function],
  "tag": "SideEffect",
}
`;

exports[`PromptToRenameActiveChart entering name of other existing chart dispatches name taken action 1`] = `
Object {
  "sideEffect": [Function],
  "tag": "SideEffect",
}
`;

exports[`PromptToRenameActiveChart entering nothing in the prompt dispatches nothing 1`] = `
Object {
  "sideEffect": [Function],
  "tag": "SideEffect",
}
`;

exports[`PromptToRenameActiveChart entering the same name as active chart in the prompt dispatches nothing 1`] = `
Object {
  "sideEffect": [Function],
  "tag": "SideEffect",
}
`;

exports[`PromptToRenameActiveChart entering unique name dispatches rename action 1`] = `
Object {
  "sideEffect": [Function],
  "tag": "SideEffect",
}
`;

exports[`RenameActiveChart 1`] = `
Object {
  "state": Object {
    "activeChartIndex": 0,
    "albumIDCounter": 3,
    "apiKey": "Test API key",
    "charts": Array [
      Object {
        "albums": Array [
          Object {
            "id": 1,
            "placeholder": true,
          },
          Object {
            "id": 2,
            "placeholder": true,
          },
          Object {
            "id": 3,
            "placeholder": true,
          },
        ],
        "name": "Renamed test chart",
        "rowsX": 10,
        "rowsY": 10,
        "shape": Object {
          "size": 40,
          "tag": "Top",
        },
      },
    ],
    "screenshot": Object {
      "loading": false,
      "scale": 2,
    },
    "search": Object {
      "query": "Test query",
      "tag": "Waiting",
    },
  },
  "tag": "Update",
}
`;

exports[`RenameActiveChart 2`] = `
Object {
  "state": Object {
    "activeChartIndex": 0,
    "albumIDCounter": 3,
    "apiKey": "Test API key",
    "charts": Array [
      Object {
        "albums": Array [
          Object {
            "id": 1,
            "placeholder": true,
          },
          Object {
            "id": 2,
            "placeholder": true,
          },
          Object {
            "id": 3,
            "placeholder": true,
          },
        ],
        "name": "Renamed test chart",
        "rowsX": 10,
        "rowsY": 10,
        "shape": Object {
          "size": 40,
          "tag": "Top",
        },
      },
      Object {
        "albums": Array [
          Object {
            "id": 4,
            "placeholder": true,
          },
          Object {
            "id": 5,
            "placeholder": true,
          },
          Object {
            "id": 6,
            "placeholder": true,
          },
        ],
        "name": "Test chart 2",
        "rowsX": 10,
        "rowsY": 10,
        "shape": Object {
          "size": 40,
          "tag": "Top",
        },
      },
    ],
    "screenshot": Object {
      "loading": false,
      "scale": 2,
    },
    "search": Object {
      "query": "Test query",
      "tag": "Waiting",
    },
  },
  "tag": "Update",
}
`;

exports[`RenameActiveChart 3`] = `
Object {
  "state": Object {
    "activeChartIndex": 2,
    "albumIDCounter": 3,
    "apiKey": "Test API key",
    "charts": Array [
      Object {
        "albums": Array [
          Object {
            "id": 1,
            "placeholder": true,
          },
          Object {
            "id": 2,
            "placeholder": true,
          },
          Object {
            "id": 3,
            "placeholder": true,
          },
        ],
        "name": "Test chart 1",
        "rowsX": 10,
        "rowsY": 10,
        "shape": Object {
          "size": 40,
          "tag": "Top",
        },
      },
      Object {
        "albums": Array [
          Object {
            "id": 4,
            "placeholder": true,
          },
          Object {
            "id": 5,
            "placeholder": true,
          },
          Object {
            "id": 6,
            "placeholder": true,
          },
        ],
        "name": "Test chart 2",
        "rowsX": 10,
        "rowsY": 10,
        "shape": Object {
          "size": 40,
          "tag": "Top",
        },
      },
      Object {
        "albums": Array [
          Object {
            "id": 7,
            "placeholder": true,
          },
          Object {
            "id": 8,
            "placeholder": true,
          },
          Object {
            "id": 9,
            "placeholder": true,
          },
        ],
        "name": "Renamed test chart",
        "rowsX": 10,
        "rowsY": 10,
        "shape": Object {
          "size": 40,
          "tag": "Top",
        },
      },
    ],
    "screenshot": Object {
      "loading": false,
      "scale": 2,
    },
    "search": Object {
      "query": "Test query",
      "tag": "Waiting",
    },
  },
  "tag": "Update",
}
`;

exports[`ShowChartNameTakenMessage calls alert 1`] = `
Object {
  "sideEffect": [Function],
  "tag": "SideEffect",
}
`;

exports[`UpdateAPIKey 1`] = `
Object {
  "state": Object {
    "activeChartIndex": 0,
    "albumIDCounter": 3,
    "apiKey": "New API key",
    "charts": Array [
      Object {
        "albums": Array [
          Object {
            "id": 1,
            "placeholder": true,
          },
          Object {
            "id": 2,
            "placeholder": true,
          },
          Object {
            "id": 3,
            "placeholder": true,
          },
        ],
        "name": "Test chart 1",
        "rowsX": 10,
        "rowsY": 10,
        "shape": Object {
          "size": 40,
          "tag": "Top",
        },
      },
    ],
    "screenshot": Object {
      "loading": false,
      "scale": 2,
    },
    "search": Object {
      "query": "Test query",
      "tag": "Waiting",
    },
  },
  "tag": "Update",
}
`;

exports[`UpdateActiveChart 1`] = `
Object {
  "state": Object {
    "activeChartIndex": 123,
    "albumIDCounter": 3,
    "apiKey": "Test API key",
    "charts": Array [
      Object {
        "albums": Array [
          Object {
            "id": 1,
            "placeholder": true,
          },
          Object {
            "id": 2,
            "placeholder": true,
          },
          Object {
            "id": 3,
            "placeholder": true,
          },
        ],
        "name": "Test chart 1",
        "rowsX": 10,
        "rowsY": 10,
        "shape": Object {
          "size": 40,
          "tag": "Top",
        },
      },
    ],
    "screenshot": Object {
      "loading": false,
      "scale": 2,
    },
    "search": Object {
      "query": "Test query",
      "tag": "Waiting",
    },
  },
  "tag": "Update",
}
`;
